toolchain/check/import_ref.cpp

```diff
@@ -879,11 +881,26 @@ class ImportRefResolver {
          .definition_id = function.definition_id.is_valid()
                               ? function_decl_id
                               : SemIR::InstId::Invalid});
+    function_decl.type_id = context_.GetFunctionType(function_decl.function_id);
     // Write the function ID into the FunctionDecl.
     context_.ReplaceInstBeforeConstantUse(function_decl_id, function_decl);
     return {context_.constant_values().Get(function_decl_id)};
   }
 
+  auto TryResolveTypedInst(SemIR::FunctionType inst) -> ResolveResult {
+    auto initial_work = work_stack_.size();
+    CARBON_CHECK(inst.type_id == SemIR::TypeId::TypeType);
+    auto fn_const_id = GetLocalConstantId(
+        import_ir_.functions().Get(inst.function_id).decl_id);
+    if (HasNewWork(initial_work)) {
+      return ResolveResult::Retry();
+    }
+    auto fn_val =
+        context_.insts().GetAs<SemIR::StructValue>(fn_const_id.inst_id());
+    CARBON_CHECK(context_.types().Is<SemIR::FunctionType>(fn_val.type_id));
+    return {context_.types().GetConstantId(fn_val.type_id)};
```

Comment:
I think it'd be simpler to import the function type constant here, rather than importing the function value constant and then grabbing its type:
```suggestion
    auto fn_const_id = GetLocalConstantId(
        import_ir_.insts()
                  .Get(import_ir_.functions().Get(inst.function_id).decl_id)
                  .type_id());
    if (HasNewWork(initial_work)) {
      return ResolveResult::Retry();
    }
    return {context_.GetTypeIdForTypeConstant(fn_val.type_id)};
```
This is fine as-is if that doesn't work or is undesirable for some reason.

Reply:
Get(function.decl_id).type_id() will correspond to the StructValue whose type is the FunctionType, creating a circular reference.

By querying the constant of the decl, failures add the decl itself to the resolve stack, which results in the type generation.

Reply:
(or maybe it's actually more direct than that -- type_id() is this FunctionType)

Reply:
Right, of course. We create the `FunctionType` as a side effect of importing the `FunctionDecl`, so we need to import the `FunctionDecl` here. Makes sense.